segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,450 @@
+            val bytesWritten = chunkStorage.write(h, systemJournalOffset, bytes.length, new ByteArrayInputStream(bytes));
+            Preconditions.checkState(bytesWritten == bytes.length);
+            systemJournalOffset += bytesWritten;
+
+            // Add a new log file if required.
+            if (!chunkStorage.supportsAppend()) {
+                currentFileIndex++;
+                systemJournalOffset = 0;
+            }
+        }
+    }
+
+    /**
+     * Formats a log record for newly added chunk.
+     *
+     * @param segmentName Name of the segment.
+     * @param offset offset at which new chunk was added.
+     * @param oldChunkName Name of the previous last chunk.
+     * @param newChunkName Name of the new last chunk.
+     * @return Formatted log record string.
+     */
+    public String getChunkAddedRecord(String segmentName, long offset, String oldChunkName, String newChunkName) {
+        return String.format("%s:%s:%s:%s:%d",
+                ADD_RECORD,
+                segmentName,
+                oldChunkName == null ? "" : oldChunkName,
+                newChunkName == null ? "" : newChunkName,
+                offset);
+    }
+
+    /**
+     * Formats a log record for trucate operation on given segment.
+     * @param segmentName Name of the segment.
+     * @param offset Offset at which the segment is truncated.
+     * @param firstChunkName Name of the new first chunk.
+     * @param startOffset Offset at which new chunk starts. The actual start offset of the segment may be anywhere in that chunk.
+     * @return Formatted log record string.
+     */
+    public String getSegmentTruncatedRecord(String segmentName, long offset, String firstChunkName, long startOffset) {
+        return String.format("%s:%s:%d:%s:%d",
+                TRUNCATE_RECORD,
+                segmentName,
+                offset,
+                firstChunkName,
+                startOffset);
+    }
+
+    /**
+     * Bootstrap the metadata about critical storage segments by reading and processing the journal.
+     *
+     * @throws Exception Exception in case of any error.
+     */
+    synchronized public void bootstrap() throws Exception {
+        synchronized (LOCK) {
+            try (val txn = metadataStore.beginTransaction()) {
+                // Step 1: Create segment metadata.
+                for (String systemSegment : systemSegments) {
+                    SegmentMetadata segmentMetadata = SegmentMetadata.builder()
+                            .name(systemSegment)
+                            .ownerEpoch(epoch)
+                            .maxRollinglength(segmentRollingPolicy.getMaxLength())
+                            .build();
+                    segmentMetadata.setActive(true).setOwnershipChanged(true);
+                    segmentMetadata.checkInvariants();
+                    txn.create(segmentMetadata);
+                    txn.markPinned(segmentMetadata);
```

Comment:
I'm not sure what the semantics of pinning a txn is.

Reply:
updated comments